### PR TITLE
[URP] [upgrader] container support

### PR DESCRIPTION
**Pull request by martint-unity**
---
### Purpose of this PR
We will need a way to structure our converters in something we call containers.
This is so that we could have 2D converters by themselves. 
Built-in to URP converters by themselves. 

converter_container

### CODE REVIEW

**Review by unity-cchu — COMMENTED** [review]
I tested this PR with the container I created, and it seems like the code is working well.

**Review by unity-cchu — APPROVED** [review]

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs [hunk 1]
```diff
@@ -117,17 +120,36 @@ internal static void DontSaveToLayout(EditorWindow wnd)
...
+            {
+                GetConverters();
+            }
+        }
+
+        void GetConverters()
+        {
+            m_CoreConvertersList.Clear();
+            m_ConverterStates.Clear();
+            m_ItemsToConvert.Clear();
```
Comment (svens-unity) [inline com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:142]: Should these not also get cleared if the` m_ConverterContainers.Any(`) check fails?

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs [hunk 2]
```diff
@@ -156,19 +178,45 @@ void OnEnable()
...
+                var button = rootVisualElement.Q<Button>("convertButton");
+                button.RegisterCallback<ClickEvent>(Convert);
+                button.SetEnabled(false);
+
+                var initButton = rootVisualElement.Q<Button>("initializeButton");
+                initButton.RegisterCallback<ClickEvent>(InitializeAllActiveConverters);
+            }
+        }
+
+        void ReCreateUI()
```
Comment (svens-unity) [inline com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:200]:
```suggestion
        void RecreateUI()
```
Recreate is one word

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs [hunk 3]
```diff
@@ -117,17 +120,36 @@ internal static void DontSaveToLayout(EditorWindow wnd)
...
+            m_ItemsToConvert.Clear();
             var converterList = TypeCache.GetTypesDerivedFrom<RenderPipelineConverter>();
 
             for (int i = 0; i < converterList.Count; ++i)
             {
-                // Iterate over the converters
+                // Iterate over the converters that are used by the current container
                 RenderPipelineConverter conv = (RenderPipelineConverter)Activator.CreateInstance(converterList[i]);
-                m_CoreConvertersList.Add(conv);
+                if (conv.container == m_ConverterContainers[m_ContainerChoiceIndex])
```
Comment (svens-unity) [inline com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:149]: Were keeping duplicate information atm if I am not mistaken. We are storing both m_CurrentContainer and m_ConverterContainers[m_ContainerChoiceIndex]. Perhaps we could just use one of the two? Maybe m_CurrentContainer.GetType() would work here.

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs [hunk 4]
```diff
@@ -117,17 +120,36 @@ internal static void DontSaveToLayout(EditorWindow wnd)
...
+            m_CoreConvertersList.Clear();
+            m_ConverterStates.Clear();
+            m_ItemsToConvert.Clear();
             var converterList = TypeCache.GetTypesDerivedFrom<RenderPipelineConverter>();
 
             for (int i = 0; i < converterList.Count; ++i)
             {
-                // Iterate over the converters
+                // Iterate over the converters that are used by the current container
                 RenderPipelineConverter conv = (RenderPipelineConverter)Activator.CreateInstance(converterList[i]);
```
Comment (svens-unity) [inline com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:148]: Another note, if we used attributes on the class we would not have to instantiate every converter that exists to check if it fits the current selected category.

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs [hunk 5]
```diff
@@ -117,17 +120,36 @@ internal static void DontSaveToLayout(EditorWindow wnd)
...
-
             // This is the drop down choices.
             m_ConverterContainers = TypeCache.GetTypesDerivedFrom<RenderPipelineConverterContainer>();
+            foreach (var cont in m_ConverterContainers)
+            {
+                var container = (RenderPipelineConverterContainer)Activator.CreateInstance(cont);
+                m_Containers.Add(container);
+                m_ContainerChoices.Add(container.name);
+            }
+
```
Comment (svens-unity) [inline com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:131]: Should `m_ContainerChoiceIndex` be reset to 0? All other code below assumes` m_ContainerChoiceIndex < m_Containers.Length`

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs [hunk 6]
```diff
@@ -156,19 +178,45 @@ void OnEnable()
...
+            m_SerializedObject.Update();
             // This is temp now to get the information filled in
-            RenderPipelineConverterContainer converterContainer =
-                (RenderPipelineConverterContainer)Activator.CreateInstance(m_ConverterContainers[0]);
-            rootVisualElement.Q<Label>("conversionName").text = converterContainer.name;
-            rootVisualElement.Q<TextElement>("conversionInfo").text = converterContainer.info;
+            rootVisualElement.Q<DropdownField>("conversionsDropDown").RegisterCallback<ChangeEvent<string>>((evt) =>
+            {
+                m_ContainerChoiceIndex = rootVisualElement.Q<DropdownField>("conversionsDropDown").index;
+                m_CurrentContainer = m_Containers[m_ContainerChoiceIndex];
```
Comment (svens-unity) [inline com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:207]: Do we need a safety check if the choice is within valid range?

**Review by svens-unity — APPROVED** [review]
Some minor comments to consider.

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs [hunk 7]
```diff
@@ -117,17 +120,36 @@ internal static void DontSaveToLayout(EditorWindow wnd)
 
         void OnEnable()
         {
-            m_CoreConvertersList = new List<RenderPipelineConverter>();
-
             // This is the drop down choices.
             m_ConverterContainers = TypeCache.GetTypesDerivedFrom<RenderPipelineConverterContainer>();
+            foreach (var cont in m_ConverterContainers)
```
[the remark this answers is not in the document]
Comment (martint-unity) [reply com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:125]: Done

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs (hunk 1 above)
Comment (martint-unity) [reply com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:142]: Done

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs (hunk 2 above)
Comment (martint-unity) [reply com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:200]: Done

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs (hunk 3 above)
Comment (martint-unity) [reply com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:149]: Unconfused this!

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs (hunk 4 above)
Comment (martint-unity) [reply com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:148]: Agreed. Added to TODO

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs (hunk 5 above)
Comment (martint-unity) [reply com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:131]: OK for now

File: com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs (hunk 6 above)
Comment (martint-unity) [reply com.unity.render-pipelines.universal/Editor/Converter/RenderPipelineConvertersEditor.cs:207]: OK for now

**Review by erikabar — APPROVED** [review]
LGTM

**Review by phi-lira — APPROVED** [review]

**Pull request merged**